app/channels/job_outputs_channel.rb

```diff
@@ -50,27 +52,28 @@ def status_response(event)
   # some kind of buffered broadcast channel would be ideal
   def subscribed
     id = params.fetch(:id)
-    if execution = JobQueue.find_by_id(id)
-      execution.viewers.push current_user
-      Thread.new do
-        ActiveRecord::Base.connection_pool.with_connection do
-          builder = EventBuilder.new(execution.job)
-          execution.output.each do |event, data|
-            transmit event: event, data: builder.payload(event, data)
-          end
-          # TODO: disconnect all listeners so they close their sockets ?
-          # then replace the reloaded/finished/waitUntilEnabled stuff with that
+
+    Thread.new do
+      ActiveRecord::Base.connection_pool.with_connection do
+        MAX_WAIT.times do
+          # running
+          execution = JobQueue.find_by_id(id)
+          break stream_execution(execution) if execution
+
+          # cancelled or we missed it running
+          job = Job.find(id)
+          break stream_finished_output(job) if job.finished?
+
+          # pending
+          sleep 1
         end
       end
-    else # job has already stopped ... send fake output (reproduce by deploying a bad ref)
-      job = Job.find(id)
-      builder = EventBuilder.new(job)
-      transmit event: :started, data: builder.payload(:started, nil)
-      job.output.each_line do |line|
-        transmit event: :message, data: builder.payload(:message, line)
-      end
-      transmit event: :finished, data: builder.payload(:finished, nil)
     end
+
+    # TODO: kill threads when user stops streaming
```

Comment:
got a branch for that ready `grosser/start-2`but it's kinda tricky since it would cancel streams when user opens multiple tabs and might have a race condition if user reloads and sends the disconnect too late .. so not doing that for now

Reply:
kk, fixed :)